clang/test/Misc/pragma-attribute-strict-subjects.c

```diff
@@ -50,8 +50,8 @@ struct testRecoverStrictnessStruct { };
 
 #pragma clang attribute pop
 
-#pragma clang attribute push (__attribute__((abi_tag("a"))), apply_to = any(function, record(unless(is_union)), variable, enum))
-// expected-error@-1 {{attribute 'abi_tag' cannot be applied to 'enum'}}
+#pragma clang attribute push (__attribute__((abi_tag("a"))), apply_to = any(function, record(unless(is_union)), variable, enum)) // expected-warning {{attribute 'abi_tag' cannot be applied to a 'void' parameter}}
+                                                                                                                                 // expected-error@-1 {{attribute 'abi_tag' cannot be applied to 'enum'}}
```

Comment:
```suggestion
#pragma clang attribute push (__attribute__((abi_tag("a"))), apply_to = any(function, record(unless(is_union)), variable, enum))
// expected-warning@-1 {{attribute 'abi_tag' cannot be applied to a 'void' parameter}}
// expected-error@-2 {{attribute 'abi_tag' cannot be applied to 'enum'}}
```
I *think* this is actually a bug with the way `clang attribute push` works. A `void` parameter is not a variable because it doesn't really declare a parameter (there's no object backing the parameter like there is for other cases). WDYT @erichkeane?

Reply:
Ah, yeah.  I think we need to fix this ASAP.  the `push` pragma needs to not consider `void` to be a `parameter` (even if it is a `ParmVarDecl` at one point).  

Whether it needs to be fixed in THIS patch, or a followup (since we jsut did the branch, we at least have some time), I'll leave up to the author.

Side note: `pragma clang attribute push` diagnostics are THE WORST here.  The fact that the error/warning is on the attribute location, but doesn't point out which declaration is causing it is awful.  We should probably find some way to intuit that we should probably have a 'note' here that ALSO points out which declaration it applies to.

Reply:
I think we can fix the behavior here by updating somewhere around https://github.com/llvm/llvm-project/blob/eb0af4e48d0e039849c6bbf36e791610e7ef9a06/clang/utils/TableGen/ClangAttrEmitter.cpp#L4473

That's what gets used by `ParsedAttr::appliesToDecl()` which is called by `Sema::AddPragmaAttributes()`.

Reply:
@AaronBallman if we're considering changes to `Attr.td`, it seems that `[Var]` and `[ParmVar]` should be extended to `AttrSubject`, since `AttrSubjectMatcherRule` only accepts a list of `AttrSubject`. I'm unsure if this qualifies as a `Subject`, though...

https://github.com/llvm/llvm-project/blob/50082773223b9eced296d8223ca4e4a79ecdeb9b/clang/include/clang/Basic/Attr.td#L560-L567

or explicitly add a helper to exclude void parameters and use it with

https://github.com/llvm/llvm-project/blob/d00579be39e8a470d7a0ff79ff6deadf9e003781/clang/lib/Sema/ParsedAttr.cpp#L172-L175

> Side note: pragma clang attribute push diagnostics are THE WORST here...

I suppose the diagnostic for void parameters will be skipped if the rules exclude them, or should the diagnostic still handle this case?

Reply:
> @AaronBallman if we're considering changes to `Attr.td`, 

I don't think we need changes in Attr.td itself; the changes should be limited to just the tablegen bits (or the part consuming the tablegen bits).

> it seems that `[Var]` and `[ParmVar]` should be extended to `AttrSubject`, since `AttrSubjectMatcherRule` only accepts a list of `AttrSubject`. I'm unsure if this qualifies as a `Subject`, though...

Those already are an `AttrSubject`; we have attributes using them already.

> https://github.com/llvm/llvm-project/blob/50082773223b9eced296d8223ca4e4a79ecdeb9b/clang/include/clang/Basic/Attr.td#L560-L567

I think that's correct as-is.

> or explicitly add a helper to exclude void parameters and use it with
> 
> https://github.com/llvm/llvm-project/blob/d00579be39e8a470d7a0ff79ff6deadf9e003781/clang/lib/Sema/ParsedAttr.cpp#L172-L175

I think that's what needs to be changed to exclude `void` parameters (or somewhere nearby that call).

> > Side note: pragma clang attribute push diagnostics are THE WORST here...
> 
> I suppose the diagnostic for void parameters will be skipped if the rules exclude them, or should the diagnostic still handle this case?

The goal is to get rid of the diagnostics about `void` parameters. e.g., `// expected-warning {{attribute 'abi_tag' cannot be applied to a 'void' parameter}}` shouldn't be fired.